### PR TITLE
Take Scala standard in spark support

### PR DESCRIPTION
**Pull request by takezoe**
This pull request includes following fix:

- Avoid procedure syntax method declaration
  - Because it brings potential of bug in the future code change. Procedure syntax is deprecated currently in the Scala standard and it will be removed in the future version of Scala.
- Avoid unnecessary `return` statement
  - `return` is translated into throwing an exception (and catching it) by Scala compiler, so we shouldn't use it in the performance reason. Also compiler can't inference a return type of the method if return is being used in the method.
- Avoid unnecessary `while` loop and mutable variable
  - Typically, we can write code safer by using immutable variables. `while` loop depends mutable flags in many cases, but it can be replaced with other way in Scala. It makes possible to eliminate mutable flags.

And I have questions about code in `ScalaValueReader.scala`, so I added inline comment to that code.

----

Thank you for submitting a pull request! 

Please make sure you have signed our [Contributor License Agreement (CLA)][].  
We are not asking you to assign copyright to us, but to give us the right to distribute your code without restriction. We ask this of all contributors in order to assure our users of the origin and continuing existence of the code.  
You only need to sign the CLA once.

- [x] I have signed the [Contributor License Agreement (CLA)][]

[Contributor License Agreement (CLA)]: https://www.elastic.co/contributor-agreement/

### CODE REVIEW

File: spark/core/main/scala/org/elasticsearch/spark/serialization/ScalaValueReader.scala [hunk 1]
```diff
@@ -71,11 +73,11 @@ class ScalaValueReader extends ValueReader with SettingsAware {
 
   def readValue(parser: Parser, value: String, esType: FieldType) = {
     if (esType == null) {
-      null
+      null // TODO Does this intend 'return null'?
```
Comment (takezoe) [inline spark/core/main/scala/org/elasticsearch/spark/serialization/ScalaValueReader.scala]: What is this null check intended? It looks no meaning...

File: spark/core/main/scala/org/elasticsearch/spark/serialization/ScalaValueReader.scala [hunk 2]
```diff
@@ -71,11 +73,11 @@ class ScalaValueReader extends ValueReader with SettingsAware {
...
     }
 
     if (parser.currentToken() == VALUE_NULL) {
-      nullValue()
+      nullValue() // TODO Does this intend 'return nullValue()'?
```
Comment (takezoe) [inline spark/core/main/scala/org/elasticsearch/spark/serialization/ScalaValueReader.scala]: Same as the above.

**Comment by jbaiera** [comment]
@takezoe thanks for the PR! I've been looking through the changes the last few days whenever I've had time. I'd like to do some digging regarding those confusing null value returns before merging, but it might take a little while with my schedule. Just wanted to pop in and give you a heads up about the status of the review. The changes look great so far!

**Comment by takezoe** [comment]
@jbaiera Thanks for your comment. I'm trying to eliminate remained return statements, but it will take a while, so I will send them as a new pull pull request after this request is merged.

File: spark/core/main/scala/org/elasticsearch/spark/serialization/ScalaValueReader.scala [hunk 3]
```diff
@@ -206,55 +208,41 @@ class ScalaValueReader extends ValueReader with SettingsAware {
...
+    _arrayDepth(potentialArray, 0)
   }
 
   def wrapArray(array: AnyRef, extraDepth: Int): AnyRef = {
       var arr = array
-      var i = 0
-      for (i <- 0 until extraDepth) {
-          arr = List(array)
+      for (_ <- 0 until extraDepth) {
+          arr = List(array) // TODO Should this line be arr = List(arr)?
```
Comment (takezoe) [inline spark/core/main/scala/org/elasticsearch/spark/serialization/ScalaValueReader.scala]: This line looks like a bug. Maybe correct code is `arr = List(arr)`?

**Comment by takezoe** [comment]
@jbaiera I pushed more fix, and add a comment to the line that looks like a bug. Hope you will check them.

**Comment by rabbitonweb** [comment]
Given the nulls might be a potential bug, maybe it is worth investigating?

**Comment by takezoe** [comment]
Hi @jbaiera, what's the status of this pull request?

**Comment by jbaiera** [comment]
Hi @takezoe, I'll be getting a proper review here very soon. My attention has been allocated toward other things at Elastic for the last month or so, but will be hopefully freeing up here in the next few days to return to here and clear out the backlog of things that need review. I seriously appreciate your patience on the matter!

**Comment by takezoe** [comment]
@jbaiera Thanks, no problem if there are nothing blocking this pull request for now.

**Comment by jbaiera** [comment]
@takezoe How did you end up tracking down all of these problems? I'm wondering if there's a way to establish a build check to make sure these sorts of problems don't arise during regular development going forward, like a code analyzer or setting the compiler to error on syntax warnings.

**Comment by takezoe** [comment]
@jbaiera In fact, I found out them by my eyes. There are some lint tools for Scala but almost of them are provided as sbt plugin, so they can't be used in Gradle.

However Scala compiler has some options to print warnings, it's might be helpful to find parts of these problems. Perhaps it's possible to enable these options in Gradle as well. I will raise an another pull request to update Gradle configuration to specify Scala compiler options.

**Comment by takezoe** [comment]
I created a new request to add Scala compiler options as #964.

**Comment by takezoe** [comment]
@jbaiera Thanks for your review. All fixed and changed the pull request target to `5.x` branch.

File: spark/sql-20/src/main/scala/org/elasticsearch/spark/sql/RowValueReader.scala [hunk 4]
```diff
@@ -52,7 +52,7 @@ private[sql] trait RowValueReader extends SettingsAware {
     }
   }
 
-  def addToBuffer(esRow: ScalaEsRow, key: AnyRef, value: Any) {
+  def addToBuffer(esRow: ScalaEsRow, key: AnyRef, value: Any): Unit = {
     val pos = esRow.rowOrder.indexOf(key.toString())
     if (pos < 0 || pos >= esRow.values.size) {
       // geo types allow fields which are ignored - need to skip these if they are not part of the schema
```
Comment (jbaiera) [inline spark/sql-20/src/main/scala/org/elasticsearch/spark/sql/RowValueReader.scala]: Thanks!

File: spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/RowValueReader.scala [hunk 5]
```diff
@@ -52,15 +52,17 @@ private[sql] trait RowValueReader extends SettingsAware {
...
-  def addToBuffer(esRow: ScalaEsRow, key: AnyRef, value: Any) {
+  def addToBuffer(esRow: ScalaEsRow, key: AnyRef, value: Any): Unit = {
     val pos = esRow.rowOrder.indexOf(key.toString())
     if (pos < 0 || pos >= esRow.values.size) {
       // geo types allow fields which are ignored - need to skip these if they are not part of the schema
       if (pos < 0 && currentFieldIsGeo) {
-        return
+        ()
+      } else {
+        throw new EsHadoopIllegalStateException(s"Position for '$sparkRowField' not found in row; typically this is caused by a mapping inconsistency")
```
Comment (jbaiera) [inline spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/RowValueReader.scala]: Thanks!

**Pull request merged**